spec/std/file_utils_spec.cr

```diff
@@ -467,4 +467,179 @@ describe "FileUtils" do
       FileUtils.rm([path1, path2, path2])
     end
   end
+
+  describe "ln" do
+    it "should create a hardlink" do
+      path1 = "/tmp/crystal_ln_test_#{Process.pid}"
+      path2 = "/tmp/crystal_ln_test_#{Process.pid + 1}"
+
+      begin
+        FileUtils.touch(path1)
+        FileUtils.ln(path1, path2)
+        File.exists?(path2).should be_true
+        File.symlink?(path2).should be_false
+      ensure
+        FileUtils.rm_rf([path1, path2])
+      end
+    end
+
+    it "should create a hardlink inside a destination dir" do
+      path1 = "/tmp/crystal_ln_test_#{Process.pid}"
+      path2 = "/tmp/crystal_ln_test_#{Process.pid + 1}/"
+      path3 = File.join(path2, File.basename(path1))
+
+      begin
+        FileUtils.touch(path1)
+        FileUtils.mkdir(path2)
+        FileUtils.ln(path1, path2)
+        File.exists?(path3).should be_true
+        File.symlink?(path3).should be_false
+      ensure
+        FileUtils.rm_rf([path1, path2])
+      end
+    end
+
+    it "should create multiple hardlinks inside a destination dir" do
+      paths = 3.times.map { |i| "/tmp/crystal_ln_test_#{Process.pid + i}" }.to_a
+      dir_path = "/tmp/crystal_ln_test_#{Process.pid + 3}/"
+
+      begin
+        paths.each { |path| FileUtils.touch(path) }
+        FileUtils.mkdir(dir_path)
+        FileUtils.ln(paths, dir_path)
+
+        paths.each do |path|
+          link_path = File.join(dir_path, File.basename(path))
+          File.exists?(link_path).should be_true
+          File.symlink?(link_path).should be_false
+        end
+      ensure
+        FileUtils.rm_rf(paths)
+        FileUtils.rm_rf(dir_path)
+      end
+    end
+
+    it "should fail with a nonexistent source" do
+      path1 = "/tmp/crystal_ln_test_#{Process.pid}"
+      path2 = "/tmp/crystal_ln_test_#{Process.pid + 1}"
+
+      expect_raises Errno do
```

Comment:
Would be nice if it could also check for the expected errno value itself.

Reply:
Does this look reasonable?

```crystal
expect_raises Errno do
  FileUtils.ln(path1, path2)
end

Errno.value.should eq(Errno::ENOENT)
```

I wasn't sure if `Errno.value` could be clobbed by other threads.

Reply:
Doesn't the errno exception contain the `ENOENT` as part of its message? If so, you can use `expect_raises(Errno, /ENOENT/)` to expect the method to raise an Errno, with a message containing the regex.

However, errno is a Thread Local variable, so your code should work too

Reply:
It looks like `ENOENT` gets stringified to "No such file or directory":

```
Expected Errno with message matching /ENOENT/, got #<Errno: Error creating link from /tmp/crystal_ln_test_28472 to /tmp/crystal_ln_test_28473: No such file or directory> with backtrace:
```

I guess I could match on that, but it feels a little less reliable. I'll go with the `Errno.value` check (I don't know how I forgot that `errno` is thread local 🙂).

Reply:
Better to use exception object for that:

```cr
ex = expect_raises Errno do
  FileUtils.ln(path1, path2)
end

ex.value.should eq(Errno::ENOENT)
```

Reply:
Thanks @Sija!